subsys/net/ip/tcp.c

```diff
@@ -1230,3 +1231,71 @@ struct net_buf *net_tcp_set_chksum(struct net_pkt *pkt, struct net_buf *frag)
 
 	return frag;
 }
+
+int tcp_parse_opts(struct net_pkt *pkt, int opt_totlen, u16_t *mss)
+{
+	struct net_buf *frag = pkt->frags;
+	u16_t pos = net_pkt_ip_hdr_len(pkt)
+		  + net_pkt_ipv6_ext_len(pkt)
+		  + sizeof(struct net_tcp_hdr);
+
+	u8_t opt, optlen;
+
+	while (opt_totlen) {
+		frag = net_frag_read(frag, pos, &pos, sizeof(opt), &opt);
+		opt_totlen--;
+
+		/* https://www.iana.org/assignments/tcp-parameters/tcp-parameters.xhtml#tcp-parameters-1 */
+		/* "Options 0 and 1 are exactly one octet which is their
+		 * kind field.  All other options have their one octet
+		 * kind field, followed by a one octet length field,
+		 * followed by length-2 octets of option data."
+		 */
+		if (opt == NET_TCP_END_OPT) {
+			break;
+		} else if (opt == NET_TCP_NOP_OPT) {
+			continue;
+		}
+
+		if (!opt_totlen) {
+			optlen = 0;
+			goto error;
+		}
+
+		frag = net_frag_read(frag, pos, &pos, sizeof(optlen), &optlen);
```

Comment:
same here

Reply:
Along the discussion on IRC, I'm adding single check at the top of function.